### ISSUE TITLE
Create a BufferedPipeReader

### ISSUE THREAD

**Issue by jkotalik** [aspnet/AspNetCore#8165]
It can be used by the form reader here: https://github.com/aspnet/AspNetCore/blob/f2e6e6ff334176540ef0b3291122e359c2106d1a/src/Http/Http/src/Features/FormFeature.cs#L139-L142. This can be a fairly simple wrapper which continually calls AdvanceTo(Start, End) on the inner PipeReader.

**Comment by jkotalik** [comment]
@davidfowl

**Comment by Tratcher** [comment]
Except then you need a way to get back to the lower level unconsumed pipe. `EnablePipeRewind` and `RewindPipe()`?

**Comment by davidfowl** [comment]
Right, there's no way to seek so maybe a new API to get the body would be better.

**Comment by anurse** [comment]
We're waiting on the corefx work to settle a bit before we can do this.

**Comment by davidfowl** [comment]
Megabacklog
